tests/test_autoscaler.py

```diff
@@ -0,0 +1,179 @@
+        autoscaler._now = Mock(return_value=1430)
+        autoscaler.scale(app)
+        mock_get_statsd_client.return_value.gauge.assert_called_once_with("{}.instance-count".format(app_name), 4)
+        mock_paas_client.return_value.assert_not_called()
+
+    def test_scale_paas_app_fewer_instances_missing_recent_scale_information(self, mock_get_statsd_client,
+                                                                             mock_paas_client, *args):
+        app_guid = '11111-11111-11111111-1111'
+        app_name = 'app-name-1'
+        cf_info = {'name': app_name, 'instances': 4, 'guid': app_guid}
+        app = self._get_mock_app(app_name, cf_info)
+        app.get_desired_instance_count = Mock(return_value=3)
+
+        autoscaler = Autoscaler()
+        autoscaler.cooldown_seconds_after_scale_up = 300
+        autoscaler.cooldown_seconds_after_scale_down = 60
+        autoscaler._now = Mock(return_value=1000)
+        autoscaler.scale(app)
+        mock_get_statsd_client.return_value.gauge.assert_called_once_with("{}.instance-count".format(app_name), 4)
+        mock_paas_client.return_value.assert_not_called()
+
+
+@patch.object(AwsBaseScaler, '_get_boto3_client')
+@patch('app.autoscaler.PaasClient')
+@patch('app.autoscaler.get_statsd_client')
+@patch.object(ScheduleScaler, '_now', return_value=datetime.datetime(2018, 3, 15, 4, 20, 00))
+@patch.object(ElbScaler, 'gauge')
+@patch.object(ElbScaler, '_get_boto3_client')
+@patch.object(ElbScaler, '_get_request_counts')
+class TestAutoscalerAlmostEndToEnd:
+    def test_scale_up(self,
+                      mock_get_request_counts,
+                      mock_get_boto3_client,
+                      mock_gauge,
+                      mock_schedule_scaler_now,
+                      mock_get_statsd_client,
+                      mock_paas_client,
+                      *args):
+        app_name = 'test-api-app'
+        app_config = {
+            'name': app_name,
+            'scalers': ['ElbScaler', 'ScheduleScaler'],
+            'elb_name': 'my-elb',
+            'min_instances': 5,
+            'max_instances': 10,
+            'threshold': 300,
+            'schedule': '''
+---
+workdays:
+  - 08:00-19:00
+weekends:
+  - 09:00-17:00
+scale_factor: 0.8
+'''
+        }
+
+        mock_paas_client.return_value.get_paas_apps.return_value = {
+            app_name: {'name': app_name, 'instances': 5, 'guid': app_name + '-guid'},
+        }
+
+        # to trigger a scale up we need at least one value greater than min_instances * threshold
+        mock_get_request_counts.return_value = [1300, 1500, 1600, 1700, 1700]
+        app_config['schedule'] = yaml.safe_load(app_config['schedule'])
+        app = App(**app_config)
+
+        autoscaler = Autoscaler()
+        autoscaler.cooldown_seconds_after_scale_up = 300
+        autoscaler.cooldown_seconds_after_scale_down = 60
+        autoscaler._now = Mock(datetime.datetime(2018, 3, 17, 10, 20, 00).timestamp())
```

Comment:
if you use freezegun you can get rid of all these scary datetime mocks https://github.com/spulec/freezegun